File: frontend/src/features/sync/api/syncV2.ts
/**
 * v2 sync API client.
 */

import api from '@/api/client';
import type { Operation } from '@/runtime';

export type VersionVector = Record<string, number>;
export type BaseVector = Record<string, VersionVector>;

export interface OperationIntentV2 {
  type: string;
  client_id: string;
  seq: number;
  node_uuid: string;
  parent_uuid?: string | null;
  after_uuid?: string | null;
  content_ast?: unknown[] | null;
  name?: string | null;
  class_uuid?: string | null;
  tag_uuid?: string | null;
  class_uuids?: string[] | null;
  tag_uuids?: string[] | null;
  is_deleted?: boolean | null;
  is_page?: boolean;
  is_task?: boolean;
  is_daily?: boolean;
  is_monthly?: boolean;
  is_yearly?: boolean;
  properties?: Record<string, unknown> | null;
  property_uuid?: string | null;
  property_value?: unknown;
}

export interface SyncBatchRequestV2 {
  ops: OperationIntentV2[];
  base_vector: BaseVector;
  workspace_uuid?: string | null;
}

export interface SyncBatchResponseV2 {
  applied: boolean;
  new_vectors: BaseVector;
}

export interface SyncConflictResponseV2 {
  stale_nodes: string[];
  server_vectors: BaseVector;
  conflict_type: 'text_edit' | 'tree_conflict' | 'permission_denied' | 'node_deleted';
}

export async function syncBatchV2(
  request: SyncBatchRequestV2,
): Promise<SyncBatchResponseV2> {
  const response = await api.post<SyncBatchResponseV2>('/sync/batch', request, {
    headers: { 'X-Notees-Sync-Protocol': 'v2' },
  });
  return response.data;
}

export function operationToIntentV2(op: Operation, clientId: string, seq: number): OperationIntentV2 | null {
  const base: OperationIntentV2 = {
    type: op.type,
    client_id: clientId,
    seq,
    node_uuid: op.blockId,
  };

  switch (op.type) {
    case 'update_content': {
      const payload = op.payload as { contentAST?: unknown[] };
      return { ...base, content_ast: payload.contentAST ?? null };
    }
    case 'update_node': {
      const payload = op.payload as { updates?: Record<string, unknown> };
      return { ...base, properties: payload.updates ?? null };
    }
    case 'create': {
      const payload = op.payload as {
        parentId?: string | null;
        afterBlockId?: string | null;
        contentAST?: unknown[];
        classIds?: string[];
        tagIds?: string[];
      };
      return {
        ...base,
        type: 'create',
        parent_uuid: payload.parentId ?? null,
        after_uuid: payload.afterBlockId ?? null,
        content_ast: payload.contentAST ?? null,
        class_uuids: payload.classIds ?? null,
        tag_uuids: payload.tagIds ?? null,
      };
    }
    case 'move':
    case 'move_node': {
      const payload = op.payload as { parentId?: string | null; afterBlockId?: string | null };
      return {
        ...base,
        type: 'move',
        parent_uuid: payload.parentId ?? null,
        after_uuid: payload.afterBlockId ?? null,
      };
    }
    case 'set_classes': {
      const payload = op.payload as { classIds?: string[] };
      return {
        ...base,
        type: 'update_node',
        properties: { class_uuids: payload.classIds ?? [] },
      };
    }
    case 'set_tags': {
      const payload = op.payload as { tagIds?: string[] };
      return {
        ...base,
        type: 'update_node',
        properties: { tag_uuids: payload.tagIds ?? [] },
      };
    }
    case 'delete':
      return { ...base, type: 'delete', is_deleted: true };
    case 'add_class': {
      const payload = op.payload as { classId?: string };
      return { ...base, type: 'add_class', class_uuid: payload.classId };
    }
    case 'remove_class': {
      const payload = op.payload as { classId?: string };
      return { ...base, type: 'remove_class', class_uuid: payload.classId };
    }
    case 'add_tag': {
      const payload = op.payload as { tagId?: string };
      return { ...base, type: 'add_tag', tag_uuid: payload.tagId };
    }
    case 'remove_tag': {
      const payload = op.payload as { tagId?: string };
      return { ...base, type: 'remove_tag', tag_uuid: payload.tagId };
    }
    default:
      return null;
  }
}

File: app/domain/entities/sync_v2.py
"""
Domain DTOs for the v2 optimistic vector-clock sync protocol.

These models define the contract between the sync server and any client
(web, Flutter, desktop). They are transport-agnostic and live in the domain
layer so the v2 sync service can be tested without FastAPI imports.
"""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field

#: Version vector for a single node: {client_id: seq}.
VersionVector = dict[str, int]

#: Base vector for a batch: {node_uuid: VersionVector}.
BaseVector = dict[str, VersionVector]


class OperationIntent(BaseModel):
    """A single client operation sent to the sync server.

    The `type` discriminator determines which payload fields are relevant.
    Unknown operation types are rejected by the server.
    """

    type: str = Field(
        ...,
        description=(
            "Operation type: create, update_content, move, delete, restore, "
            "add_class, remove_class, add_tag, remove_tag, update_node, set_property."
        ),
    )
    client_id: str = Field(..., description="Stable client/device UUID.")
    seq: int = Field(..., description="Monotonic client sequence number for this op.")
    node_uuid: str = Field(..., description="UUID of the target node.")
    # Tree / hierarchy
    parent_uuid: str | None = Field(None, description="New parent UUID for tree ops.")
    after_uuid: str | None = Field(None, description="Sibling to insert/move after.")
    # Content
    content_ast: list[dict[str, Any]] | None = Field(
        None, description="Full AST document for update_content ops."
    )
    name: str | None = Field(None, description="Plain-text name/title for update_node ops.")
    # Classes / tags
    class_uuid: str | None = Field(None, description="Class UUID for add_class/remove_class.")
    tag_uuid: str | None = Field(None, description="Tag UUID for add_tag/remove_tag.")
    class_uuids: list[str] | None = Field(None, description="Complete class list for set_classes.")
    tag_uuids: list[str] | None = Field(None, description="Complete tag list for set_tags.")
    # Node flags
    is_deleted: bool | None = Field(None, description="Soft-delete flag for delete/restore ops.")
    # Extra node properties (icon, color, etc.)
    properties: dict[str, Any] | None = Field(
        None, description="Arbitrary node property updates for update_node ops."
    )
    # Property value ops
    property_uuid: str | None = Field(
        None, description="Property UUID for set_property ops."
    )
    property_value: Any | None = Field(
        None, description="Property value for set_property ops."
    )
    # Node type flags for create ops
    is_page: bool = Field(False, description="Create as a page.")
    is_task: bool = Field(False, description="Create as a task item.")
    is_daily: bool = Field(False, description="Create as a daily journal page.")
    is_monthly: bool = Field(False, description="Create as a monthly journal page.")
    is_yearly: bool = Field(False, description="Create as a yearly journal page.")


class SyncBatchRequest(BaseModel):
    """Request body for POST /sync/batch."""

    ops: list[OperationIntent] = Field(
        ..., description="Client operations to apply.", max_length=100
    )
    base_vector: BaseVector = Field(
        ...,
        description=(
            "Last server-confirmed version vector per affected node. "
            "Must NOT include optimistic local increments."
        ),
    )
    workspace_uuid: str | None = Field(
        None, description="Target workspace UUID; defaults to the user's active workspace."
    )


class SyncBatchResponse(BaseModel):
    """Successful response from POST /sync/batch."""

    applied: bool = True
    new_vectors: BaseVector = Field(
        ..., description="Updated version vectors for all nodes touched by the batch."
    )


class SyncConflictResponse(BaseModel):
    """409 response from POST /sync/batch when one or more nodes are stale."""

    stale_nodes: list[str] = Field(
        ..., description="Node UUIDs whose base vector did not match the server."
    )
    server_vectors: BaseVector = Field(
        ..., description="Current server vectors for the stale nodes."
    )
    conflict_type: str = Field(
        ...,
        description=(
            "Broad conflict category: text_edit, tree_conflict, "
            "permission_denied, node_deleted."
        ),
    )


class AppliedOperation(BaseModel):
    """An operation that was applied on the server, broadcast to other clients."""

    type: str
    node_uuid: str
    client_id: str
    seq: int
    parent_uuid: str | None = None
    after_uuid: str | None = None
